sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala

```diff
@@ -207,32 +230,47 @@ case class GlobalLimitExec(limit: Int, child: SparkPlan) extends BaseLimitExec {
  * the child's single output partition.
  */
 case class GlobalLimitAndOffsetExec(
-    limit: Int,
+    limit: Int = -1,
     offset: Int,
     child: SparkPlan) extends BaseLimitExec {
+  assert(offset > 0)
 
   override def requiredChildDistribution: List[Distribution] = AllTuples :: Nil
 
-  override def doExecute(): RDD[InternalRow] = child.execute().mapPartitions { iter =>
-    iter.take(limit + offset).drop(offset)
+  override def doExecute(): RDD[InternalRow] = if (limit >= 0) {
+    child.execute().mapPartitions(iter => iter.take(limit + offset).drop(offset))
```

Comment:
not related to this PR, but since we are touch here, let's use `mapPartitionsInternal` which is more efficient.

Reply:
OK